staging/src/k8s.io/client-go/tools/portforward/websocket-dialer.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package portforward
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/transport/websocket"
+	"k8s.io/klog/v2"
+)
+
+// wsDialer implements "httpstream.Dial" interface
+type wsDialer struct {
+	method    string
+	url       *url.URL
+	transport http.RoundTripper
+	holder    websocket.ConnectionHolder
+	ports     []string
+}
+
+// NewWebSocketDialer creates and returns the wsDialer structure which implemements the "httpstream.Dialer"
+// interface. The dialer can upgrade a websocket request, creating a websocket connection. This function
+// returns an error if one occurs.
+func NewWebSocketDialer(method string, url *url.URL, config *restclient.Config, ports []string) (httpstream.Dialer, error) {
+	if len(ports) == 0 {
```

Comment:
we can validate here too , to make this limit only for websockets https://github.com/kubernetes/kubernetes/pull/120889/files#r1337841125